enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+---
+title: control-plane-machineset-vertical-scaling
+authors:
+  - "@bergmannf"
+reviewers:
+  - @"JoelSpeed"
+approvers:
+  - @"JoelSpeed"
+api-approvers: 
+  - None
+creation-date: 2023-10-30
+last-updated: 2023-10-30
+tracking-link:
+  - https://issues.redhat.com/browse/OSD-15261
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Control plane machineset automatic vertical scaling
+
+## Summary
+
+To allow control plane nodes to remain at an adequate size for its cluster, this
+proposal introduces new configuration for the control plane machineset operator
+to allow it to make automated scaling decisions for control plane node sizing.
+
+## Motivation
+
+During normal operation of an OpenShift cluster, as worker node count and
+workloads increase, it becomes necessary to increase resources available to the
+control plane nodes.
+
+During SRE operations of the managed OpenShift product, there was already a
+trigger indentified, when these increases are required. Instead of requiring the
+adjustments to be performed by hand, the operator should be able to
+automatically determine if a control plane node size change is required.
+
+On the other hand, it should also be possible to determine that control plane
+nodes have become too large for a cluster, because load has decreased. In this
+case the operator should also be able to automatically decide to scale the
+control plane down again.
+
+With the automatic increase of the control plane size, cluster performance can
+remain high, while automated decrease of the size will ensure that spend for
+users in a cloud environment is only has high as needed.
+
+### User Stories
+
+* As an OpenShift administrator, I want to know that the platform will always
+  have enough resources available to it's control plane nodes without manual
+  intervention.
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
+* Allow users to configure the sizes that can be chosen for automatic scaling.
+* Allow users to configure the thresholds when scaling will occur for CPU and
+  memory usage.
+* Allow users to configure how often checks for automatic scaling should be run.
+* Allow users to configure how long after a scale up or scale down, no more
+  scaling should be performed.
+* Allow users to completely disable scale up.
+* Allow users to completely disable scale down.
+
+### Non-Goals
+
+* Horizontal scaling of control plane nodes
```

Comment:
I'd be curious, could you expand why this isn't a goal? Would horizontal scaling solve the problem or is it completely out of the question?

Reply:
Yeah I think that would be a case for an alternative.

I think horizontal would solve the problems here, but you're right, it's currently untested. But this does seem like a good change to revisit whether it should become supported or not.